src/delta.c

```diff
@@ -53,10 +60,90 @@ void __create_delta(struct file *file, int from_version, char *from_hash)
+		 * That xattr contains a hash of the file content. We know that
+		 * the file content has changed, so  the xattr will be different
+		 * from the one on the old file and we can bail out early without
+		 * even bothering with retrieving the original file. A better
+		 * solution for systems with IMA would be to support deltas even
+		 * when xattrs are different.
+		 */
+		goto out;
+	}
+
+	if (access(original, F_OK) &&
+	    content_url) {
+		/* File does not exist. Try to get it from the online update repo instead.
+		 * This fallback is meant to be used for CI builds which start with no local
+		 * state and only HTTP(S) access to the published www directory.
+		 * Not being able to retrieve the file is not an error and will merely
+		 * prevent computing the delta.
+		 */
+		string_or_die(&tmpdir, "%s/make-pack-tmpdir-XXXXXX", state_dir);
+		tmpdir = g_dir_make_tmp("make-pack-XXXXXX", &gerror);
+		if (!tmpdir) {
+			LOG(NULL, "Failed to create temporary directory for untarring original file", "%s",
+			    gerror->message);
+			assert(0);
+		}
+		/* Determine hash of original file in the corresponding Manifest. */
+		manifest = manifest_from_file(from_version, "full");
+		if (!manifest) {
+			LOG(NULL, "Failed to read full Manifest", "version %d, cannot retrieve original file",
+			    from_version);
+			goto out;
+		}
+		const char *last_hash = NULL;
+		GList *list = g_list_first(manifest->files);
+		while (list) {
+			struct file *original_file = list->data;
+			if (!strcmp(file->filename, original_file->filename)) {
+				last_hash = original_file->hash;
+				break;
+			}
+			list = g_list_next(list);
+		}
+		if (!last_hash) {
+			LOG(NULL, "Original file not found", "%s in full manifest for %d - inconsistent update data?!",
+			    file->filename, from_version);
+			goto out;
+		}
+
+		/* We use a temporary copy because we don't want to
+		 * tamper with the original "full" folder which
+		 * probably does not even exist. Using a temporary file
+		 * file implies re-downloading in the future, but that's
+		 * consistent with the intended usage in a CI environment
+		 * which always starts from scratch.
+		 */
+		free(original);
+		string_or_die(&original, "%s/%s", tmpdir, last_hash);
+		delete_original = true;
+
+		/*
+		 * This is a proof-of-concept. A real implementation should use
+		 * a combination of libcurl + libarchive calls to unpack the files.
+		 * For current Ostro OS, deltas despite xattr differences would
+		 * be needed, otherwise this code here is of little use (all
+		 * modified files fail the xattr sameness check, because security.ima
+		 * changes when file content changes).
+		 */
+		string_or_die(&url, "%s/%d/files/%s.tar", content_url, from_version, last_hash);
+		LOG(file, "Downloading original file", "%s to %s", url, original);
+
+		/* bsdtar can detect compression when reading from stdin, GNU tar can't. */
+		string_or_die(&cmd, "curl -s %s | bsdtar -C %s -xf -", url, tmpdir);
```

Comment:
@pohly would you please update this PR to use libcurl/libarchive like you mention in your comment? We will likely be accepting your PR to introduce libarchive into swupd-server once https://github.com/clearlinux/swupd-client/pull/327 has been merged and #48 can be tested again.